tick/optim/model/src/model_labels_features.cpp

```diff
@@ -1,21 +1,39 @@
-// License: BSD 3 clause
-
-//
-// Created by Stéphane GAIFFAS on 06/12/2015.
-//
-
 #include "model_labels_features.h"
 
+// License: BSD 3 clause
+
 ModelLabelsFeatures::ModelLabelsFeatures(SBaseArrayDouble2dPtr features,
                                          SArrayDoublePtr labels)
     : n_samples(labels.get() ? labels->size() : 0),
       n_features(features.get() ? features->n_cols() : 0),
       labels(labels),
-      features(features) {
+      features(features),
+      ready_columns_sparsity(false) {
   if (labels.get() && labels->size() != features->n_rows()) {
     std::stringstream ss;
     ss << "In ModelLabelsFeatures, number of labels is " << labels->size();
     ss << " while the features matrix has " << features->n_rows() << " rows.";
     throw std::invalid_argument(ss.str());
   }
 }
+
+void ModelLabelsFeatures::compute_columns_non_zeros(ArrayULong &out_columns_non_zeros) {
+  if (features->is_sparse()) {
+    if (out_columns_non_zeros.size() != n_features) {
+      TICK_ERROR("given `out_columns_non_zeros` vector must match `n_features`")
+    }
+    out_columns_non_zeros.fill(0.);
+    for (ulong i = 0; i < n_samples; ++i) {
+      BaseArrayDouble features_i = get_features(i);
+      for (ulong j = 0; j < features_i.size_sparse(); ++j) {
+        // If the entry is indeed non-zero (nothing forbids to store zeros...) increment
+        // the number of non-zeros of the columns
+        if (features_i.data()[j] != 0) {
+          out_columns_non_zeros[features_i.indices()[j]] += 1;
```

Comment:
Just a quick thought,
If an entry in a sparse matrix is 0 and we discard it here we should take it into account in SVRG code as well... Probably is it not such a good idea to discard it here...

Reply:
Indeed it leads to false results

```python
    def test_dense_and_sparse_match_fake_non_zeros(self):
        """...Test in SVRG that dense and sparse code matches in all possible
        settings
        """
        variance_reductions = ['last', 'rand']
        rand_types = ['unif', 'perm']
        seed = 123
        tol = 0.
        max_iter = 20
        X, y = self.simu_linreg_data(interc=-1.)
        d = X.shape[1]

        proxs = [
            ProxTV(strength=1e-2, range=(0, d)),
            ProxElasticNet(strength=1e-2, ratio=0.9, range=(0, d)),
            ProxEquality(range=(0, d))
        ]

        # Let's create model manually
        X_sparse = csr_matrix(X)
        # fill half of the non zeros with zeros
        n_new_zeros = int(len(X_sparse.data) / 2)
        mask = np.random.choice(len(X_sparse.data), n_new_zeros, replace=False)
        # next line makes SVRG fail...
        X_sparse.data[mask] = 0
        X_dense = X_sparse.toarray()
        model_dense = ModelLinReg().fit(X_dense, y)
        model_spars = ModelLinReg().fit(X_sparse, y)

        step = 1 / model_spars.get_lip_max()
        for variance_reduction, rand_type, prox in product(variance_reductions,
                                                           rand_types, proxs):
            solver_sparse = SVRG(step=step, tol=tol, max_iter=max_iter,
                                 verbose=False,
                                 variance_reduction=variance_reduction,
                                 rand_type=rand_type, seed=seed) \
                .set_model(model_spars) \
                .set_prox(prox)

            solver_dense = SVRG(step=step, tol=tol, max_iter=max_iter,
                                verbose=False,
                                variance_reduction=variance_reduction,
                                rand_type=rand_type, seed=seed) \
                .set_model(model_dense) \
                .set_prox(prox)

            solver_sparse.solve()
            solver_dense.solve()
            np.testing.assert_array_almost_equal(solver_sparse.solution,
                                                 solver_dense.solution, 7)
```

Reply:
Nice catch ! I'll remove the test then